### PR TITLE
Gl/intf

### PR DESCRIPTION
**Pull request by guoqilu**

### CODE REVIEW

File: rst_html_wrapper.template [hunk 1]
```diff
@@ -1,4 +1,4 @@
 {TITLE}
-{TILE_UNDERSZAP}
-.. raw: html
-   :file: {HTML_FILE}
\ No newline at end of file
+{TITLE_UNDERSCORE}
```
Comment (ciglass) [inline rst_html_wrapper.template:2]: Why did you need to change the number of `:` characters after `raw`?

File: tests/golden_inputs/BUILD [hunk 2]
```diff
@@ -17,7 +17,7 @@ golden_pkg_tests(
 
 golden_intf_tests(
     deps = {
-        "test_intf_a": [
```
Comment (ciglass) [inline tests/golden_inputs/BUILD]: Why did you rename the `test_intf_a`? I wanted to keep the `_a` identifier for when we want to add other test interfaces.

File: yis_gen.py [hunk 3]
```diff
@@ -2170,6 +2194,12 @@ def computed_width(self):
             return self.width.computed_width
         return self.sv_type.computed_width
 
+    def computed_port_name(self, connection):
```
Comment (ciglass) [inline yis_gen.py:2184]: I think this style would be more pythonic:

```suggestion
    def computed_port_name(self, connection):
      src = self.parent.block
      dst = connection['name']
      if self.direction == "input":
        src = connection['name']
        dst = self.parent.block
      return f"{src}__{dst}_{connection['submodule']}_{self.name}"
```

I think it's more typical python style to assign defaults, then check a conditional for an overwrite. I think it's more common in Perl (and probably other languages) to do a single line with multiple conditions.

File: tests/golden_inputs/test_intf.yis [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# -*- yaml -*-
```
Comment (ciglass) [inline tests/golden_inputs/test_intf.yis]: Shouldn't there be changes to the associated test_intf.html?

File: tests/golden_inputs/test_intf.yis [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        submodules: # [optional] if multiple interface instances to same other block, defines a submodule name to differentiate
+          # list the names of variants of this interface instantied as.
+          #   each submodule will make a port named <src>__<dst>__<submodule>_<symbol> where submodule will be name of submodule
+          - name: hero # min 2 lowercase letters/numbers, no _ allowed.
+          - name: ace
+      - name: foo
+    symbols: # defines symbols of this interface
+      - name: write # min 2 lowercase letters/numbers, no _ allowed.
+        type: test_pkg_a::hero_write_t
+        direction: input  # [optional] if omitted, default to be an input
```
Comment (ciglass) [inline tests/golden_inputs/test_intf.yis]: This comment says the default should be an input, but I don't think we should allow an implicit direction.

**Review by ciglass — CHANGES_REQUESTED** [review]

File: README.md [hunk 6]
```diff
@@ -93,19 +93,26 @@ An intf uses the primitives defined in pkgs to describe the interface between tw
 An intf consists of one or more components, and each component consists of one or more connections.  
 
 components are groups of individual port-to-port connections that constitute a higher-level protocol or logical grouping of signals.
-For example, if two blocks had a val/rdy handshake, the "val" port and the "rdy" port would each be described in a connection, and those connections would be contained within a component.
+Ports describe the shape of an interface (component) and connections describe the repeat of an interface (component).
```
Comment (ciglass) [inline README.md]: I understand what you're going for with 'shape' and 'repeat', but I think it would be clearer to say something like 'Ports describe the specific SV-types and widths that make up a component, and connections describe if the component is instantiated multiple times on the interface".

File: README.md [hunk 7]
```diff
@@ -93,19 +93,26 @@ An intf uses the primitives defined in pkgs to describe the interface between tw
...
+For example, if two blocks had a `val/rdy` handshake, the `val` and `rdy` are ports described in `ports` section.
+if this same handshake ports are used multiply to connecting other blocks, those are described inthe `connections` section
```
Comment (ciglass) [inline README.md]: Can you take another pass through your modified section of the readme to ensure that all sentences start with a capital letter (unless the term is surrounded by `ticks` for specific terms) and end with a period? I know it can be annoying some times to figure out where sentence boundaries are in markdown. It might be easier to render the markdown as HTML then use that to review.

File: rst_html_wrapper.template (hunk 1 above)
Comment (guoqilu) [reply rst_html_wrapper.template:2]: otherwise it breaks. I compared with the template in proj_doc_server and found this

File: yis_gen.py (hunk 3 above)
Comment (guoqilu) [reply yis_gen.py:2184]: I'm fine with that, is there a way I can directly take your suggestion? or do I have to change code and re-push?

File: tests/golden_inputs/test_intf.yis (hunk 5 above)
Comment (guoqilu) [reply tests/golden_inputs/test_intf.yis]: I'm fine with that, we can take the default out and make it required field

File: yis_gen.py (hunk 3 above)
Comment (ciglass) [reply yis_gen.py:2184]: I think there is a way to do it automatically, but you can copy the code out manually since there are only a few lines.

File: README.md (hunk 6 above)
Comment (guoqilu) [reply README.md]: I said similar below at each section's description. meant to avoid repeat of words that's why I looked for a "vague" word to replace.

File: README.md (hunk 7 above)
Comment (guoqilu) [reply README.md]: sure. I wish renderer could automatically capitalize for me...

File: tests/golden_inputs/BUILD (hunk 2 above)
Comment (guoqilu) [reply tests/golden_inputs/BUILD]: can we do `test_a_intf` instead then, I meant to have `_intf_yis` and `_intf.yis` being standing Bazel target name and file name convention for consistency when people creating `.yis` files. having `intf` in between is a bit harder to do

File: yis_gen.py (hunk 3 above)
Comment (guoqilu) [reply yis_gen.py:2184]: done

File: README.md (hunk 6 above)
Comment (guoqilu) [reply README.md]: updated

File: README.md (hunk 7 above)
Comment (guoqilu) [reply README.md]: done

File: tests/golden_inputs/BUILD (hunk 2 above)
Comment (guoqilu) [reply tests/golden_inputs/BUILD]: nvm, figured out, pushed up

File: tests/golden_inputs/test_intf.yis (hunk 4 above)
Comment (guoqilu) [reply tests/golden_inputs/test_intf.yis]: added

File: tests/golden_inputs/BUILD (hunk 2 above)
Comment (ciglass) [reply tests/golden_inputs/BUILD]: This is a good point. On the next PR, we should revisit the naming conventions and explicit naming requirements both on packages and interfaces.

File: README.md [hunk 8]
```diff
@@ -93,19 +93,26 @@ An intf uses the primitives defined in pkgs to describe the interface between tw
 An intf consists of one or more components, and each component consists of one or more connections.  
 
 components are groups of individual port-to-port connections that constitute a higher-level protocol or logical grouping of signals.
-For example, if two blocks had a val/rdy handshake, the "val" port and the "rdy" port would each be described in a connection, and those connections would be contained within a component.
+Ports describe how an interface (component) looks like in terms of width and direction.
+Connections describe how if an interface (component) instantiated multiple times.
```
Comment (ciglass) [inline README.md]:
```suggestion
Connections describe if an interface (component) is instantiated multiple times.
```

File: README.md [hunk 9]
```diff
@@ -93,19 +93,26 @@ An intf uses the primitives defined in pkgs to describe the interface between tw
 An intf consists of one or more components, and each component consists of one or more connections.  
 
 components are groups of individual port-to-port connections that constitute a higher-level protocol or logical grouping of signals.
-For example, if two blocks had a val/rdy handshake, the "val" port and the "rdy" port would each be described in a connection, and those connections would be contained within a component.
+Ports describe how an interface (component) looks like in terms of width and direction.
```
Comment (ciglass) [inline README.md]: Why is 'component' used in parenthesis throughout this README? I think you can describe what a component is relative to an interface once near the top of the `yis_intf` section, then reference 'component' later on in the docs without spelling out the full 'interface (component)'

File: README.md (hunk 8 above)
Comment (guoqilu) [reply README.md]: updated

File: README.md (hunk 9 above)
Comment (guoqilu) [reply README.md]: updated

**Review by ciglass — APPROVED** [review]
